### PR TITLE
Catch exception in TryConvertTo [updated]

### PR DESCRIPTION
**Pull request by SteveSandersonMS**
This is a continuation of https://github.com/dotnet/aspnetcore/pull/33594. I had to create a new PR because the original one was no longer mergeable and further tweaks were needed.

### CODE REVIEW

File: src/Components/Components/src/BindConverter.cs [hunk 1]
```diff
@@ -1933,6 +1962,14 @@ internal static class ParserDelegateCache
                 {
                     parser = ConvertToNullableTimeOnly;
                 }
+                else if (typeof(T) == typeof(Guid))
+                {
+                    parser = ConvertToGuid;
+                }
+                else if (typeof(T) == typeof(Guid?))
```
Comment (pranavkm) [inline src/Components/Components/src/BindConverter.cs:1969]: Kinda unrelated - I imagine this block of code would look a lot nicer as a switch expression.

File: src/Components/Components/test/BindConverterTest.cs [hunk 2]
```diff
@@ -298,6 +298,77 @@ public void FormatValue_TypeConverter()
         Assert.Equal(expected, actual);
     }
 
+    [Fact]
+    public void TryConvertTo_Guid_Valid()
+    {
+        // Arrange
+        var expected = Guid.NewGuid();
```
Comment (pranavkm) [inline src/Components/Components/test/BindConverterTest.cs:305]: Super nit, if this fails for some arbitrary input, we would never find out what the input value was. Maybe we should invert  lines 312 and 313 or just use a fixed Guid.

**Review by pranavkm — APPROVED** [review]

File: src/Components/Components/test/BindConverterTest.cs [hunk 3]
```diff
@@ -298,6 +298,77 @@ public void FormatValue_TypeConverter()
...
+        var incomingValue = expected.ToString();
+
+        // Act
+        var successfullyConverted = BindConverter.TryConvertTo<Guid>(incomingValue, CultureInfo.CurrentCulture, out var actual);
+
+        // Assert
+        Assert.True(successfullyConverted);
+        Assert.Equal(expected, actual);
```
Comment (SteveSandersonMS) [inline src/Components/Components/test/BindConverterTest.cs:313]:
```suggestion
        Assert.Equal(expected, actual);
        Assert.True(successfullyConverted);
```

File: src/Components/Components/test/BindConverterTest.cs (hunk 2 above)
Comment (SteveSandersonMS) [reply src/Components/Components/test/BindConverterTest.cs:305]: Good point, though I'll be pretty astonished if some GUIDs work and others don't. Swapped the two lines around as suggested anyway :)

File: src/Components/Components/src/BindConverter.cs (hunk 1 above)
Comment (SteveSandersonMS) [reply src/Components/Components/src/BindConverter.cs:1969]: Yes, probably. This code predates switch expressions. Keeping this out-of-scope for the PR though.

**Pull request merged**